File: apple/packages/EnteComponents/Sources/EnteComponents/IconAction.swift
import SwiftUI

public struct IconAction<Content: View>: View {
    private let kind: IconActionKind
    private let enabled: Bool
    private let size: CGFloat
    private let action: () -> Void
    private let content: Content

    public init(
        kind: IconActionKind = .unfilled,
        enabled: Bool = true,
        size: CGFloat = 40,
        action: @escaping () -> Void,
        @ViewBuilder content: () -> Content
    ) {
        self.kind = kind
        self.enabled = enabled
        self.size = size
        self.action = action
        self.content = content()
    }

    public var body: some View {
        Button(action: action) {
            content
                .frame(width: size, height: size)
                .contentShape(Rectangle())
        }
        .buttonStyle(
            IconActionStyle(
                kind: kind,
                cornerRadius: kind == .circular ? size / 2 : EnteRadius.medium
            )
        )
        .disabled(!enabled)
    }
}

public enum IconActionKind: Sendable {
    case primary
    case unfilled
    case accent
    case circular
    case onMedia
}

private struct IconActionStyle: ButtonStyle {
    @Environment(\.entePalette) private var palette
    @Environment(\.isEnabled) private var enabled
    let kind: IconActionKind
    let cornerRadius: CGFloat

    func makeBody(configuration: Configuration) -> some View {
        let colors = colors(pressed: configuration.isPressed)
        return configuration.label
            .foregroundStyle(colors.foreground)
            .background(colors.background)
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius, style: .continuous))
            .scaleEffect(configuration.isPressed ? 0.98 : 1)
            .frame(minWidth: 44, minHeight: 44)
            .contentShape(Rectangle())
            .animation(.easeOut(duration: EnteMotion.quick), value: configuration.isPressed)
    }

    private func colors(pressed: Bool) -> (background: Color, foreground: Color) {
        if !enabled, kind != .onMedia {
            return (kind == .unfilled ? .clear : palette.fill, palette.hintText)
        }
        return switch kind {
        case .primary, .circular: (pressed ? palette.fillDarker : palette.surface, palette.text)
        case .unfilled: (.clear, palette.text.opacity(palette.isDark ? 1 : 0.75))
        case .accent: (pressed ? palette.primaryDarker : palette.primary, .white)
        case .onMedia:
            (
                enabled && pressed ? .white.opacity(0.12) : .clear,
                .white.opacity(enabled ? 1 : 0.4)
            )
        }
    }
}

File: android/packages/components/src/main/kotlin/io/ente/components/IconAction.kt
package io.ente.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

@Composable
public fun IconAction(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    kind: IconActionKind = IconActionKind.Unfilled,
    enabled: Boolean = true,
    size: Dp = 40.dp,
    content: @Composable () -> Unit,
) {
    val palette = LocalEntePalette.current
    val interactionSource = remember { MutableInteractionSource() }
    val pressed by interactionSource.collectIsPressedAsState()
    val colors = iconColors(kind, palette, enabled, pressed)
    val scale by
        animateFloatAsState(
            targetValue = if (pressed && enabled) 0.98f else 1f,
            animationSpec = tween(EnteMotion.quick),
            label = "iconActionScale",
        )

    Box(
        modifier =
            modifier
                .sizeIn(minWidth = 48.dp, minHeight = 48.dp)
                .clickable(
                    enabled = enabled,
                    role = Role.Button,
                    interactionSource = interactionSource,
                    indication = null,
                    onClick = onClick,
                ),
        contentAlignment = Alignment.Center,
    ) {
        CompositionLocalProvider(LocalContentColor provides colors.foreground) {
            Box(
                modifier =
                    Modifier.size(size)
                        .graphicsLayer {
                            scaleX = scale
                            scaleY = scale
                        }
                        .background(
                            colors.background,
                            RoundedCornerShape(
                                if (kind == IconActionKind.Circular) size / 2 else EnteRadius.medium
                            ),
                        ),
                contentAlignment = Alignment.Center,
            ) {
                content()
            }
        }
    }
}

public enum class IconActionKind {
    Primary,
    Unfilled,
    Accent,
    Circular,
    OnMedia,
}

private data class IconColors(val background: Color, val foreground: Color)

private fun iconColors(
    kind: IconActionKind,
    palette: Palette,
    enabled: Boolean,
    pressed: Boolean,
): IconColors {
    if (!enabled && kind != IconActionKind.OnMedia) {
        return IconColors(
            if (kind == IconActionKind.Unfilled) Color.Transparent else palette.fill,
            palette.hintText,
        )
    }
    return when (kind) {
        IconActionKind.Primary,
        IconActionKind.Circular ->
            IconColors(if (pressed) palette.fillDarker else palette.surface, palette.text)
        IconActionKind.Unfilled ->
            IconColors(
                Color.Transparent,
                if (palette.isDark) palette.text else palette.text.copy(alpha = 0.75f),
            )
        IconActionKind.Accent ->
            IconColors(if (pressed) palette.primaryDarker else palette.primary, Color.White)
        IconActionKind.OnMedia ->
            IconColors(
                if (enabled && pressed) Color.White.copy(alpha = 0.12f) else Color.Transparent,
                Color.White.copy(alpha = if (enabled) 1f else 0.4f),
            )
    }
}
